// js-crypto-aes/dist/webapi.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.encrypt = encrypt;
exports.decrypt = decrypt;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

/**
 * webapi.js
 */

/**
 * Encrypt data through AES of WebCrypto API.
 * @param {Uint8Array} msg - Plaintext message to be encrypted.
 * @param {Uint8Array} key - Byte array of symmetric key.
 * @param {String} name - Name of AES algorithm like 'AES-GCM'.
 * @param {Uint8Array} [iv] - Byte array of initial vector if required.
 * @param {Uint8Array} [additionalData] - Byte array of additional data if required.
 * @param {Number} [tagLength] - Authentication tag length if required.
 * @param {Object} webCrypto - WebCrypto object, i.e., window.crypto.subtle or window.msCrypto.subtle
 * @return {Promise<Uint8Array>} - Encrypted data byte array.
 * @throws {Error} - Throws if UnsupportedCipher.
 */
function encrypt(_x, _x2, _x3, _x4) {
  return _encrypt.apply(this, arguments);
}
/**
 * Decrypt data through AES of WebCrypto API.
 * @param {Uint8Array} data - Encrypted message to be decrypted.
 * @param {Uint8Array} key - Byte array of symmetric key.
 * @param {String} name - Name of AES algorithm like 'AES-GCM'.
 * @param {Uint8Array} [iv] - Byte array of initial vector if required.
 * @param {Uint8Array} [additionalData] - Byte array of additional data if required.
 * @param {Number} [tagLength] - Authentication tag length if required.
 * @param {Object} webCrypto - WebCrypto object, i.e., window.crypto.subtle or window.msCrypto.subtle
 * @return {Promise<Uint8Array>} - Decrypted plaintext message.
 * @throws {Error} - Throws if UnsupportedCipher or DecryptionFailure.
 */


function _encrypt() {
  _encrypt = (0, _asyncToGenerator2.default)(
  /*#__PURE__*/
  _regenerator.default.mark(function _callee(msg, key, _ref, webCrypto) {
    var _ref$name, name, iv, additionalData, tagLength, encryptionConfig, sessionKeyObj, data, _sessionKeyObj, encryptedObj, _data;

    return _regenerator.default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            _ref$name = _ref.name, name = _ref$name === void 0 ? 'AES-GCM' : _ref$name, iv = _ref.iv, additionalData = _ref.additionalData, tagLength = _ref.tagLength;
            encryptionConfig = setCipherParams({
              name: name,
              iv: iv,
              additionalData: additionalData,
              tagLength: tagLength
            });

            if (!(typeof window.msCrypto === 'undefined')) {
              _context.next = 12;
              break;
            }

            _context.next = 5;
            return webCrypto.importKey('raw', key, encryptionConfig, false, ['encrypt', 'decrypt']);

          case 5:
            sessionKeyObj = _context.sent;
            _context.next = 8;
            return webCrypto.encrypt(encryptionConfig, sessionKeyObj, msg);

          case 8:
            data = _context.sent;
            return _context.abrupt("return", new Uint8Array(data));

          case 12:
            _context.next = 14;
            return msImportKey('raw', key, encryptionConfig, false, ['encrypt', 'decrypt'], webCrypto);

          case 14:
            _sessionKeyObj = _context.sent;
            _context.next = 17;
            return msEncrypt(encryptionConfig, _sessionKeyObj, msg, webCrypto);

          case 17:
            encryptedObj = _context.sent;

            if (!(name === 'AES-GCM')) {
              _context.next = 25;
              break;
            }

            _data = new Uint8Array(encryptedObj.ciphertext.byteLength + encryptedObj.tag.byteLength);

            _data.set(new Uint8Array(encryptedObj.ciphertext));

            _data.set(new Uint8Array(encryptedObj.tag), encryptedObj.ciphertext.byteLength);

            return _context.abrupt("return", _data);

          case 25:
            return _context.abrupt("return", new Uint8Array(encryptedObj));

          case 26:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));
  return _encrypt.apply(this, arguments);
}

function decrypt(_x5, _x6, _x7, _x8) {
  return _decrypt.apply(this, arguments);
}
/**
 * Set params for encryption algorithms.
 * @param {String} name - Name of AES algorithm like 'AES-GCM'.
 * @param {Uint8Array} [iv] - Byte array of initial vector if required.
 * @param {Uint8Array} [additionalData] - Byte array of additional data if required.
 * @param {Number} [tagLength] - Authentication tag length if required.
 */


function _decrypt() {
  _decrypt = (0, _asyncToGenerator2.default)(
  /*#__PURE__*/
  _regenerator.default.mark(function _callee2(data, key, _ref2, webCrypto) {
    var name, iv, additionalData, tagLength, decryptionConfig, sessionKeyObj, msg, _sessionKeyObj2, ciphertext, tag, _msg, _msg2;

    return _regenerator.default.wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            name = _ref2.name, iv = _ref2.iv, additionalData = _ref2.additionalData, tagLength = _ref2.tagLength;
            decryptionConfig = setCipherParams({
              name: name,
              iv: iv,
              additionalData: additionalData,
              tagLength: tagLength
            });

            if (window.msCrypto) {
              _context2.next = 12;
              break;
            }

            _context2.next = 5;
            return webCrypto.importKey('raw', key, decryptionConfig, false, ['encrypt', 'decrypt']);

          case 5:
            sessionKeyObj = _context2.sent;
            _context2.next = 8;
            return webCrypto.decrypt(decryptionConfig, sessionKeyObj, data).catch(function (e) {
              throw new Error("DecryptionFailure: ".concat(e.message));
            });

          case 8:
            msg = _context2.sent;
            return _context2.abrupt("return", new Uint8Array(msg));

          case 12:
            _context2.next = 14;
            return msImportKey('raw', key, decryptionConfig, false, ['encrypt', 'decrypt'], webCrypto);

          case 14:
            _sessionKeyObj2 = _context2.sent;

            if (!(name === 'AES-GCM')) {
              _context2.next = 24;
              break;
            }

            ciphertext = data.slice(0, data.length - tagLength);
            tag = data.slice(data.length - tagLength, data.length);
            _context2.next = 20;
            return msDecrypt(Object.assign(decryptionConfig, {
              tag: tag
            }), _sessionKeyObj2, ciphertext, webCrypto).catch(function (e) {
              throw new Error("DecryptionFailure: ".concat(e.message));
            });

          case 20:
            _msg = _context2.sent;
            return _context2.abrupt("return", new Uint8Array(_msg));

          case 24:
            _context2.next = 26;
            return msDecrypt(decryptionConfig, _sessionKeyObj2, data, webCrypto).catch(function (e) {
              throw new Error("DecryptionFailure: ".concat(e.message));
            });

          case 26:
            _msg2 = _context2.sent;
            return _context2.abrupt("return", new Uint8Array(_msg2));

          case 28:
          case "end":
            return _context2.stop();
        }
      }
    }, _callee2);
  }));
  return _decrypt.apply(this, arguments);
}

var setCipherParams = function setCipherParams(_ref3) {
  var name = _ref3.name,
      iv = _ref3.iv,
      additionalData = _ref3.additionalData,
      tagLength = _ref3.tagLength;
  var alg = {};

  switch (name) {
    case 'AES-GCM':
      {
        Object.assign(alg, {
          name: name,
          iv: iv,
          tagLength: tagLength * 8
        });
        Object.assign(alg, additionalData.length > 0 ? {
          additionalData: additionalData
        } : {});
        break;
      }

    case 'AES-CBC':
      {
        alg.name = name;
        alg.iv = iv;
        break;
      }
  }

  return alg;
}; // function definitions for IE


var msImportKey = function msImportKey(type, key, alg, ext, use, webCrypto) {
  return new Promise(function (resolve, reject) {
    var op = webCrypto.importKey(type, key, alg, ext, use);

    op.oncomplete = function (evt) {
      resolve(evt.target.result);
    };

    op.onerror = function () {
      reject('KeyImportingFailed');
    };
  });
};

var msEncrypt = function msEncrypt(alg, key, msg, webCrypto) {
  return new Promise(function (resolve, reject) {
    var op = webCrypto.encrypt(alg, key, msg);

    op.oncomplete = function (evt) {
      resolve(evt.target.result);
    };

    op.onerror = function () {
      reject('EncryptionFailure');
    };
  });
};

var msDecrypt = function msDecrypt(alg, key, data, webCrypto) {
  return new Promise(function (resolve, reject) {
    var op = webCrypto.decrypt(alg, key, data);

    op.oncomplete = function (evt) {
      resolve(evt.target.result);
    };

    op.onerror = function () {
      reject('DecryptionFailure');
    };
  });
};
